src/backend/renderer/utils.rs

```diff
@@ -19,31 +18,72 @@ use wayland_server::protocol::{wl_buffer::WlBuffer, wl_surface::WlSurface};
 
 #[derive(Default)]
 pub(crate) struct SurfaceState {
+    pub(crate) commit_count: usize,
     pub(crate) buffer_dimensions: Option<Size<i32, Buffer>>,
     pub(crate) buffer_scale: i32,
     pub(crate) buffer_transform: Transform,
     pub(crate) buffer: Option<WlBuffer>,
+    pub(crate) damage: VecDeque<Vec<Rectangle<i32, Buffer>>>,
+    pub(crate) renderer_seen: HashMap<(TypeId, usize), usize>,
     pub(crate) textures: HashMap<(TypeId, usize), Box<dyn std::any::Any>>,
     #[cfg(feature = "desktop")]
-    pub(crate) damage_seen: HashSet<crate::desktop::space::SpaceOutputHash>,
+    pub(crate) space_seen: HashMap<crate::desktop::space::SpaceOutputHash, usize>,
 }
 
+const MAX_DAMAGE: usize = 4;
+
 impl SurfaceState {
     pub fn update_buffer(&mut self, attrs: &mut SurfaceAttributes) {
         match attrs.buffer.take() {
             Some(BufferAssignment::NewBuffer { buffer, .. }) => {
                 // new contents
                 self.buffer_dimensions = buffer_dimensions(&buffer);
+
+                #[cfg(feature = "desktop")]
+                if self.buffer_scale != attrs.buffer_scale
+                    || self.buffer_transform != attrs.buffer_transform.into()
+                {
+                    self.reset_space_damage();
+                }
                 self.buffer_scale = attrs.buffer_scale;
                 self.buffer_transform = attrs.buffer_transform.into();
+
                 if let Some(old_buffer) = std::mem::replace(&mut self.buffer, Some(buffer)) {
                     if &old_buffer != self.buffer.as_ref().unwrap() {
                         old_buffer.release();
                     }
                 }
                 self.textures.clear();
-                #[cfg(feature = "desktop")]
-                self.damage_seen.clear();
+                let (new_commit_count, overflow) = self.commit_count.overflowing_add(1);
+                self.commit_count = new_commit_count;
+                if overflow {
+                    // we need to avoid bugs in `damage_since` (see below)
+                    self.damage.clear();
+                } else {
+                    self.damage.push_front({
```

Comment:
Maybe use a variable for the calculated damage to reduce the nesting as it may be hard to read.